src/Features/Core/Portable/DisposeAnalysis/DisposableFieldsShouldBeDisposedDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,266 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Concurrent;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using Analyzer.Utilities;
+using Microsoft.CodeAnalysis.CodeQuality;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.FlowAnalysis;
+using Microsoft.CodeAnalysis.FlowAnalysis.DataFlow;
+using Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.DisposeAnalysis;
+using Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.PointsToAnalysis;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.DisposeAnalysis
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    internal sealed class DisposableFieldsShouldBeDisposedDiagnosticAnalyzer
+        : AbstractCodeQualityDiagnosticAnalyzer
+    {
+        private static readonly DiagnosticDescriptor s_disposableFieldsShouldBeDisposedRule = CreateDescriptor(
+            IDEDiagnosticIds.DisposableFieldsShouldBeDisposedDiagnosticId,
+            title: new LocalizableResourceString(nameof(FeaturesResources.Disposable_fields_should_be_disposed), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+            messageFormat: new LocalizableResourceString(nameof(FeaturesResources.Disposable_field_0_is_never_disposed), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+            description: new LocalizableResourceString(nameof(FeaturesResources.DisposableFieldsShouldBeDisposedDescription), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+            isUnneccessary: false);
+
+        public DisposableFieldsShouldBeDisposedDiagnosticAnalyzer()
+            : base(ImmutableArray.Create(s_disposableFieldsShouldBeDisposedRule), GeneratedCodeAnalysisFlags.Analyze)
```

Comment:
what usual supposed to do if generated file has this issue? is it also marked as concurrent run?

Reply:
I added generated code analysis for this analyzer based on this bug filed by a user: https://github.com/dotnet/roslyn-analyzers/issues/2306

Reply:
ah. is it for intra analysis?

Reply:
Not even inter-procedural, but just regular analysis. If a disposable type defines disposable fields in non-generated code and has its Dispose method in generated code, we need to get call backs for generated code to avoid flagging fields disposed in Dispose()